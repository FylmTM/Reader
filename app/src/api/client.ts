type Response<R> =
  | { status: "ok"; payload: R }
  | { status: "error"; message: string };

function post<T, R>(path: string, data: T): Promise<R> {
  return fetch(path, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(data),
  })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

function get<R>(path: string): Promise<R> {
  return fetch(path, { method: "GET" })
    .then(response => {
      return response.json() as Promise<Response<R>>;
    })
    .then(result => {
      switch (result.status) {
        case "error":
          throw Error(result.message);
        case "ok":
          return result.payload;
      }
    });
}

export default {
  get,
  post,
};
