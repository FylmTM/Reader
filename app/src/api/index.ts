import { CategoriesWithFeeds, Post, User } from "../domain";

const DELAY = 200;

function getCurrentUser(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function login(): Promise<User> {
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve({
        id: 1,
        username: "me"
      });
    }, DELAY);
  });
}

function logout(): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function getCategoriesWithFeeds(): Promise<CategoriesWithFeeds> {
  return new Promise<CategoriesWithFeeds>((resolve, reject) => {
    setTimeout(() => {
      resolve([
        {
          category: {
            id: 1,
            name: "General"
          },
          feeds: [
            {
              id: 1,
              kind: "RSS",
              title: "Hacker News",
              link: "https://news.ycombinator.com/",
              feed: "https://news.ycombinator.com/rss"
            },
            {
              id: 2,
              kind: "RSS",
              title: "vc.ru",
              link: "https://vc.ru/",
              feed: "https://vc.ru/rss/all"
            }
          ]
        },
        {
          category: {
            id: 2,
            name: "Gaming - Best Gaming Category in the whole World"
          },
          feeds: [
            {
              id: 3,
              kind: "RSS",
              title: "StopGame - Best Gaming Website in the World",
              link: "https://stopgame.ru",
              feed: "https://rss.stopgame.ru/rss_all.xml"
            }
          ]
        }
      ]);
    }, DELAY);
  });
}

function randomInt(min: number, max: number) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getPosts(): Promise<Array<Post>> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve(
        Array(15)
          .fill(".")
          .map((_, i) => {
            return {
              id: i + 1,
              is_read: false,
              is_read_later: false,
              feed: {
                id: 1,
                category_id: 1,
                title: "Hacker News"
              },
              link: "http://example.com",
              title:
                randomInt(0, 1) === 1
                  ? "World is finally living on mars."
                  : "World is falling down to the works of ancient civilization who is going to do a revolution.",
              summary:
                randomInt(0, 1) === 1
                  ? Array(randomInt(1, 100))
                      .fill("lorem")
                      .join(" ")
                  : undefined,
              date: `${randomInt(1, 10)}h`,
              media:
                randomInt(0, 1) === 1
                  ? {
                      type: "image/jpeg",
                      // link: "http://localhost:3000/android-chrome-512x512.png"
                      link:
                        "https://stopga.me/images/video/2019/12/07/M8MFv_nN.jpg"
                    }
                  : undefined,
              comments_link:
                randomInt(0, 1) === 1
                  ? "http://example.com/#comments"
                  : undefined
            };
          })
      );
    }, DELAY);
  });
}

function markAsRead(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

function markAsReadLater(postId: number, isRead: Boolean): Promise<void> {
  return new Promise(resolve => {
    setTimeout(() => {
      resolve();
    }, DELAY);
  });
}

export default {
  getCurrentUser,
  login,
  logout,
  getCategoriesWithFeeds,
  getPosts,
  markAsRead,
  markAsReadLater
};
