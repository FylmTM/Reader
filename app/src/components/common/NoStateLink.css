span.link:hover {
    text-decoration: underline;
}
