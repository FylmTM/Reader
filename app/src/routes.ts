export const ALL_ROUTE = "/";

export const CATEGORY_ROUTE = "/category/:categoryId";
export const CATEGORY_POST_ROUTE = "/category/:categoryId/post/:postId";
export const FEED_ROUTE = "/feed/:feedId";
export const FEED_POST_ROUTE = "/feed/:feedId/post/:postId";
