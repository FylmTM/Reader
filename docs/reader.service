[Unit]
Description=Reader
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=1
User=root
Environment=ROCKET_PORT=80
Environment=RUST_LOG=warn
WorkingDirectory=/root/reader
ExecStart=/root/reader/reader

[Install]
WantedBy=multi-user.target
