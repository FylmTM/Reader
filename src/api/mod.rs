use crate::error::Error;
use rocket_contrib::json::Json;

pub mod authentication;
pub mod catchers;
pub mod endpoints;

#[derive(Serialize, Deserialize)]
pub struct ApiResponse<T> {
    status: String,
    payload: T,
}

impl<T> ApiResponse<T> {
    fn with(payload: T) -> ApiResponse<T> {
        ApiResponse {
            status: "ok".to_string(),
            payload,
        }
    }
}

type Response<T> = Result<Json<ApiResponse<T>>, Error>;

fn ok<T>(data: T) -> Response<T> {
    Ok(Json(ApiResponse::with(data)))
}
