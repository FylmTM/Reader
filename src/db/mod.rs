pub use rusqlite::{Connection, NO_PARAMS};

pub use entities::*;
pub use pool::{Pool, PoolConnection};

use crate::error::{ApplicationError, Error, Result};

pub mod bootstrap;
pub mod entities;
pub mod pool;

pub type UserId = i64;
pub type CategoryId = i64;
pub type FeedId = i64;
pub type PostId = i64;
pub type UserPostId = i64;

pub trait Queries {
    fn find_user_by_api_key(&self, api_key: &str) -> Result<User>;
    fn find_user_ids_by_feed(&self, feed_id: FeedId) -> Result<Vec<UserId>>;
    fn find_categories_by_user_id(&self, id: UserId) -> Result<Vec<Category>>;
    fn find_feeds(&self) -> Result<Vec<Feed>>;
    fn find_feeds_by_category_id(&self, category_id: CategoryId) -> Result<Vec<Feed>>;
    fn find_feed_by_id(&self, id: FeedId) -> Result<Feed>;
    fn count_posts_by_user(&self, id: UserId) -> Result<i64>;
    fn count_posts_by_category(&self, id: CategoryId) -> Result<i64>;
    fn count_posts_by_feed(&self, id: FeedId) -> Result<i64>;
    fn save_post(&self, feed_id: FeedId, post: &Post) -> Result<PostId>;
    fn save_user_post(&self, user_id: UserId, post_id: PostId) -> Result<UserPostId>;
}

impl Queries for Connection {
    fn find_user_by_api_key(self: &Connection, api_key: &str) -> Result<User> {
        // language=SQLite
        let query = "
            select u.id, u.username
            from users u
            where u.api_key = ?
        ";
        let user = self.query_row(query, &[api_key], |row| {
            Ok(User {
                id: row.get(0)?,
                username: row.get(1)?,
            })
        })?;
        Ok(user)
    }

    fn find_user_ids_by_feed(self: &Connection, feed_id: FeedId) -> Result<Vec<i64>> {
        // language=SQLite
        let query = "
            select u.id
            from users u
            inner join categories c on u.id = c.user_id
            inner join category_feeds cf on c.id = cf.category_id
            where cf.feed_id = ?
        ";
        let mut statement = self.prepare(query)?;
        let user_ids_rows = statement.query_map(&[feed_id], |row| Ok(row.get(0)?))?;

        let mut user_ids = Vec::new();
        for user_id in user_ids_rows {
            user_ids.push(user_id?);
        }
        Ok(user_ids)
    }

    fn find_categories_by_user_id(self: &Connection, id: UserId) -> Result<Vec<Category>> {
        // language=SQLite
        let query = "
            select c.id, c.name
            from categories c
            where c.user_id = ?
        ";
        let mut statement = self.prepare(query)?;
        let categories_rows = statement.query_map(&[id], |row| {
            Ok(Category {
                id: row.get(0)?,
                name: row.get(1)?,
            })
        })?;

        let mut categories = Vec::new();
        for category in categories_rows {
            categories.push(category?)
        }
        Ok(categories)
    }

    //noinspection DuplicatedCode
    fn find_feeds(self: &Connection) -> Result<Vec<Feed>> {
        // language=SQLite
        let query = "
            select f.id, f.kind, f.title, f.link, f.feed
            from feeds f
        ";
        let mut statement = self.prepare(query)?;
        let feeds_rows = statement.query_map(NO_PARAMS, map_row_to_feed)?;

        let mut feeds = Vec::new();
        for feed in feeds_rows {
            feeds.push(feed?)
        }
        Ok(feeds)
    }

    fn find_feeds_by_category_id(self: &Connection, category_id: CategoryId) -> Result<Vec<Feed>> {
        // language=SQLite
        let query = "
            select f.id, f.kind, f.title, f.link, f.feed
            from category_feeds cf
            left join feeds f on cf.feed_id = f.id
            where cf.category_id = ?
        ";
        let mut statement = self.prepare(query)?;
        let feeds_rows = statement.query_map(&[category_id], map_row_to_feed)?;

        let mut feeds = Vec::new();
        for feed in feeds_rows {
            feeds.push(feed?)
        }
        Ok(feeds)
    }

    //noinspection DuplicatedCode
    fn find_feed_by_id(self: &Connection, id: FeedId) -> Result<Feed> {
        // language=SQLite
        let query = "
            select f.id, f.kind, f.title, f.link, f.feed
            from feeds f
            where f.id = ?
        ";
        let feed = self.query_row(query, &[id], |row| {
            Ok(Feed {
                id: row.get(0)?,
                kind: row.get(1)?,
                title: row.get(2)?,
                link: row.get(3)?,
                feed: row.get(4)?,
            })
        })?;
        Ok(feed)
    }

    fn count_posts_by_user(self: &Connection, id: UserId) -> Result<i64> {
        // language=SQLite
        let query = "
            select count(up.post_id)
            from user_posts up
            where up.user_id = ?
        ";
        let count = self.query_row(query, &[id], |row| row.get(0))?;
        Ok(count)
    }

    fn count_posts_by_category(self: &Connection, id: CategoryId) -> Result<i64> {
        // language=SQLite
        let query = "
            select count(p.id)
            from categories c
            inner join category_feeds cf on c.id = cf.category_id
            inner join posts p on cf.feed_id = p.feed_id
            where c.id = ?
        ";
        let count = self.query_row(query, &[id], |row| row.get(0))?;
        Ok(count)
    }

    fn count_posts_by_feed(self: &Connection, id: FeedId) -> Result<i64> {
        // language=SQLite
        let query = "
            select count(p.id)
            from posts p
            where p.feed_id = ?
        ";
        let count = self.query_row(query, &[id], |row| row.get(0))?;
        Ok(count)
    }

    fn save_post(self: &Connection, feed_id: FeedId, post: &Post) -> Result<PostId> {
        // language=SQLite
        let query = "
            insert or ignore into posts(feed_id, link, title, date, summary, content, media_type, media_link, comments_link)
            values (:feed_id, :link, :title, :date, :summary, :content, :media_type, :media_link, :comments_link)
        ";
        let count = self.execute_named(
            query,
            &[
                (":feed_id", &feed_id),
                (":link", &post.link),
                (":title", &post.title),
                (":date", &post.date),
                (":summary", &post.summary),
                (":content", &post.content),
                (":media_type", &post.media_link),
                (":comments_link", &post.comments_link),
            ],
        )?;
        check_one_entity_inserted(count)?;
        Ok(self.last_insert_rowid())
    }

    fn save_user_post(&self, user_id: UserId, post_id: PostId) -> Result<UserPostId> {
        // language=SQLite
        let query = "
            insert or ignore into user_posts(user_id, post_id)
            values (:user_id, :post_id)
        ";
        let count = self.execute_named(query, &[(":user_id", &user_id), (":post_id", &post_id)])?;
        check_one_entity_inserted(count)?;
        Ok(self.last_insert_rowid())
    }
}

fn map_row_to_feed(row: &rusqlite::Row) -> rusqlite::Result<Feed> {
    Ok(Feed {
        id: row.get(0)?,
        kind: row.get(1)?,
        title: row.get(2)?,
        link: row.get(3)?,
        feed: row.get(4)?,
    })
}

fn check_one_entity_inserted(count: usize) -> Result<()> {
    if count == 1 {
        Ok(())
    } else if count == 0 {
        Err(Error::App(ApplicationError::QueryEntityAlreadyExists))
    } else {
        Err(Error::App(ApplicationError::QueryUnexpectedRowCount {
            expected: 1,
            actual: count as i64,
        }))
    }
}
