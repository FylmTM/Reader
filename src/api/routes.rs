use std::ops::Add;

use rocket::http::{Cookie, Cookies};
use rocket_contrib::json::Json;

use crate::api::{ok, Response};
use crate::db::{self, Queries};

#[derive(Serialize, Deserialize)]
pub struct ApiKey {
    pub api_key: String,
}

#[post("/api/v1/auth/login", data = "<api_key>")]
pub fn auth_login(
    api_key: Json<ApiKey>,
    conn: db::PoolConnection,
    mut cookies: Cookies,
) -> Response<db::User> {
    let api_key = api_key.0.api_key;
    let user = conn.find_user_by_api_key(&api_key)?;
    cookies.add(
        Cookie::build("api_key", api_key)
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::days(365)))
            .finish(),
    );

    ok(user)
}

#[post("/api/v1/auth/logout")]
pub fn auth_logout(mut cookies: Cookies) -> Response<()> {
    cookies.add(
        Cookie::build("api_key", "")
            .path("/")
            .http_only(true)
            .expires(time::now().add(time::Duration::milliseconds(1)))
            .finish(),
    );

    ok(())
}

#[get("/api/v1/current_user")]
pub fn current_user(user: db::User) -> Response<db::User> {
    ok(user)
}

#[get("/api/v1/categories_with_feeds")]
pub fn categories_with_feeds(
    user: db::User,
    conn: db::PoolConnection,
) -> Response<db::CategoriesWithFeeds> {
    let categories = conn.find_categories_by_user_id(user.id)?;

    // todo: consider optimizing by doing one query to get all feeds,
    //       and then mapping them to their categories
    //       Will it actually be faster?
    let mut categories_with_feeds = Vec::with_capacity(categories.len());
    for category in categories {
        let feeds = conn.find_feeds_by_category_id(category.id)?;
        categories_with_feeds.push(db::CategoryWithFeeds { category, feeds })
    }

    ok(categories_with_feeds)
}
