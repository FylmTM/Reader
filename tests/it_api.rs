#![feature(proc_macro_hygiene, decl_macro)]
#[macro_use]
extern crate rocket;

#[macro_use]
extern crate insta;

use rocket::uri;

use common::client::{self, ClientOperations, RequestOperations, ResponseOperations};
use reader::api::routes::*;
use reader::db;

pub mod common;

#[test]
fn test_auth_login_failed() {
    let client = client::get();
    let mut response = client
        .post_uri(uri!(auth_login))
        .json(ApiKey {
            api_key: "invalid".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity())
}

#[test]
fn test_auth_login_success() {
    let client = client::get();
    let mut response = client
        .post_uri(uri!(auth_login))
        .json(ApiKey {
            api_key: "api_key".to_string(),
        })
        .dispatch();

    assert_json_snapshot!(response.json_entity());
    assert_eq!(response.cookies().len(), 1);
}

#[test]
fn test_auth_logout_success() {
    let client = client::authenticated();

    let mut response = client.post_uri(uri!(auth_logout)).dispatch();
    assert_eq!(response.entity(), client::api_ok(()));
    assert_eq!(response.cookies().len(), 1);

    let mut response = client.get_uri(uri!(current_user)).dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user_failed_if_not_authenticated() {
    let client = client::get();

    let mut response = client.get_uri(uri!(current_user)).dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_current_user() {
    let client = client::authenticated();

    let mut response = client.get_uri(uri!(current_user)).dispatch();
    assert_json_snapshot!(response.json_entity());
}

#[test]
fn test_get_categories_with_feeds() {
    let client = client::authenticated();

    let mut response = client.get_uri(uri!(categories_with_feeds)).dispatch();
    assert_json_snapshot!(response.json_entity());
}
