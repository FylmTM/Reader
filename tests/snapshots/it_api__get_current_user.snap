---
source: tests/it_api.rs
expression: response.entity_value()
---
[
  "200 OK",
  {
    "payload": {
      "id": 1,
      "username": "me"
    },
    "status": "ok"
  }
]
